<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-GB" xml:lang="en-GB"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.29">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>2&nbsp; Methods: normalise, identify DRBs and save as SummarizedExperiment objects – Inferring the MoA of drugs based on barcode fingerprints.</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./results.html" rel="next">
<link href="./data-management.html" rel="prev">
<link href="./assets/favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-bcbce66894bfa8e7e94e40e0186b9bb8.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark-4e7a6ecc11b04d2331dcee725084a9de.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-bcbce66894bfa8e7e94e40e0186b9bb8.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-b8d8e7bbfbdf5c0c9eec33f85154acaa.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark-befac119a4bdef22a0431cf315bc3f4a.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-b8d8e7bbfbdf5c0c9eec33f85154acaa.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<script async="" src="https://hypothes.is/embed.js"></script><script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script><script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
<meta property="og:title" content="2&nbsp; Methods: normalise, identify DRBs and save as SummarizedExperiment objects – Inferring the MoA of drugs based on barcode fingerprints.">
<meta property="og:description" content="">
<meta property="og:site_name" content="Inferring the MoA of drugs based on barcode fingerprints.">
</head>
<body class="nav-sidebar docked quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./methods.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Methods: normalise, identify DRBs and save as <code>SummarizedExperiment</code> objects</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Cell Barcoding and Drug Fingerprints</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <a href="./Inferring-the-MoA-of-drugs-based-on-barcode-fingerprints..docx" title="Download Docx" class="quarto-navigation-tool px-1" aria-label="Download Docx"><i class="bi bi-file-word"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-management.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Data management: metadata correction and cleansing</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./methods.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Methods: normalise, identify DRBs and save as <code>SummarizedExperiment</code> objects</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./results.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Results: Heatmaps and Drug networks, based on barcode fingerprints</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./perspectives.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Future perspectives: single-cell integration.</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cell-lines-databases.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Appendix: publish to standardised cell line repositories.</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bibliographic References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#reproducibility" id="toc-reproducibility" class="nav-link active" data-scroll-target="#reproducibility"><span class="header-section-number">2.1</span> Reproducibility</a></li>
  <li>
<a href="#sec-batch-processing" id="toc-sec-batch-processing" class="nav-link" data-scroll-target="#sec-batch-processing"><span class="header-section-number">2.2</span> Preprocessing barcode counts</a>
  <ul>
<li><a href="#background-noise-removal-and-summarizedexperiment-stantardised-storage" id="toc-background-noise-removal-and-summarizedexperiment-stantardised-storage" class="nav-link" data-scroll-target="#background-noise-removal-and-summarizedexperiment-stantardised-storage"><span class="header-section-number">2.2.1</span> <strong>Background Noise Removal</strong> and <code>SummarizedExperiment</code> stantardised storage</a></li>
  <li><a href="#normalisation" id="toc-normalisation" class="nav-link" data-scroll-target="#normalisation"><span class="header-section-number">2.2.2</span> <strong>Normalisation</strong></a></li>
  <li><a href="#prepare-summarizedexperiment-objects-for-normalised-data" id="toc-prepare-summarizedexperiment-objects-for-normalised-data" class="nav-link" data-scroll-target="#prepare-summarizedexperiment-objects-for-normalised-data"><span class="header-section-number">2.2.3</span> Prepare <code>SummarizedExperiment</code> objects for normalised data</a></li>
  </ul>
</li>
  <li>
<a href="#differential-analyses-and-binarisation" id="toc-differential-analyses-and-binarisation" class="nav-link" data-scroll-target="#differential-analyses-and-binarisation"><span class="header-section-number">2.3</span> <strong>Differential analyses</strong> and Binarisation</a>
  <ul>
<li><a href="#summarizedexperiment-of-binarised-markers" id="toc-summarizedexperiment-of-binarised-markers" class="nav-link" data-scroll-target="#summarizedexperiment-of-binarised-markers"><span class="header-section-number">2.3.1</span> <code>SummarizedExperiment</code> of binarised markers</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/edit/master/methods.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/blob/master/methods.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content column-body" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Methods: normalise, identify DRBs and save as <code>SummarizedExperiment</code> objects</span>
</h1>
</div>



<div class="quarto-title-meta column-body">

    
  
    
  </div>
  


</header><section id="reproducibility" class="level2" data-number="2.1"><h2 data-number="2.1" class="anchored" data-anchor-id="reproducibility">
<span class="header-section-number">2.1</span> Reproducibility</h2>
<p>I list below the R packages required to reproduce the analyses.</p>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## data wrangling</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org">readxl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ycphs.github.io/openxlsx/index.html">openxlsx</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## reporting</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ardata-fr.github.io/flextable-book/">flextable</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># automated package linting</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://xml2.r-lib.org">xml2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://downlit.r-lib.org/">downlit</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## auxiliary functions</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="st">"R/utils.R"</span><span class="op">)</span></span>
<span><span class="va">today_date</span> <span class="op">&lt;-</span> <span class="st">"2025-04-28"</span></span>
<span><span class="co">## today_date &lt;- format(Sys.Date(), "%Y-%m-%d")</span></span>
<span></span>
<span><span class="co">## set the seed, for fixing generation of ComplexHeatmaps (dendogram clustering)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section><section id="sec-batch-processing" class="level2" data-number="2.2"><h2 data-number="2.2" class="anchored" data-anchor-id="sec-batch-processing">
<span class="header-section-number">2.2</span> Preprocessing barcode counts</h2>
<p><strong>Code colour</strong>: <span data-fg="blue" style="color: blue;">Luca’s protocol</span>, and <span data-fg="red" style="color: red;">Bastien’s comments and perspectives</span></p>
<section id="background-noise-removal-and-summarizedexperiment-stantardised-storage" class="level3" data-number="2.2.1"><h3 data-number="2.2.1" class="anchored" data-anchor-id="background-noise-removal-and-summarizedexperiment-stantardised-storage">
<span class="header-section-number">2.2.1</span> <strong>Background Noise Removal</strong> and <code>SummarizedExperiment</code> stantardised storage</h3>
<p><span data-fg="blue" style="color: blue;">Eliminate barcodes for which the combined counts of the 4 controls per barcode are below a given threshold, here <code>5</code>.</span></p>
<div class="cell" data-layout-align="center">
<div id="lst-background-filtering" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-background-filtering-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.1: Remove background noise.
</figcaption><div aria-describedby="lst-background-filtering-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">barcode_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"./data/barcode-counts/"</span>,</span>
<span>                            pattern <span class="op">=</span> <span class="st">"exp.*\\.csv"</span>,</span>
<span>                            full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">thresh_background</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span><span class="co">## ./data/pheno_data_metadata_2025-04-18.csv</span></span>
<span><span class="va">pheno_data</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./data/pheno_data_metadata_"</span>,</span>
<span>                                     <span class="va">today_date</span>, <span class="st">".csv"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## filename &lt;- grep( "exp200921", barcode_files, value = TRUE)[2]</span></span>
<span><span class="va">barcode_counts_aggregated</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">barcode_files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">filename</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">barcode_counts</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">filename</span><span class="op">)</span></span>
<span>  <span class="va">experience_name</span> <span class="op">&lt;-</span> <span class="va">filename</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">tools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/tools/fileutils.html">file_path_sans_ext</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## 1) explicitly change replicates name to enable 1-1 mapping ----</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">experience_name</span><span class="op">==</span><span class="st">"exp200921_dose response osim"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">barcode_counts</span> <span class="op">&lt;-</span> <span class="va">barcode_counts</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename_with</a></span><span class="op">(</span></span>
<span>        .fn <span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"exp200921_"</span>, <span class="st">"exp200921_dose_response_"</span>, <span class="va">.x</span><span class="op">)</span>,</span>
<span>        .cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="st">"Contro"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">replicates_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">barcode_counts</span><span class="op">)</span>, <span class="st">"barcode_id"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## 2) Remove lowly expressed barcodes ----</span></span>
<span>  <span class="va">control_index</span> <span class="op">&lt;-</span> <span class="va">pheno_data</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Batch_ID</span><span class="op">==</span><span class="va">experience_name</span> <span class="op">&amp;</span> </span>
<span>                    <span class="va">Compound</span><span class="op">==</span><span class="st">"Control"</span> <span class="op">&amp;</span></span>
<span>                    <span class="va">Replicates_ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">replicates_names</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">Replicates_ID</span><span class="op">)</span></span>
<span>  <span class="va">signicant_barcodes_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">barcode_counts</span><span class="op">[</span>, <span class="va">control_index</span><span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">thresh_background</span><span class="op">)</span></span>
<span>  <span class="va">filtered_barcode_counts</span> <span class="op">&lt;-</span> <span class="va">barcode_counts</span><span class="op">[</span><span class="va">signicant_barcodes_index</span>,<span class="op">]</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"\n\nWe are considering experiment:"</span>, <span class="va">experience_name</span>, <span class="st">"with"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">filtered_barcode_counts</span><span class="op">)</span>, <span class="st">"barcode IDs kept.\n\n"</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">filtered_barcode_counts</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/reduce.html">reduce</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">inner_join</a></span><span class="op">(</span><span class="va">.x</span>, <span class="va">.y</span>, by <span class="op">=</span> <span class="st">"barcode_id"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Deal with specific duplicated colnames, resulting from control duplicates</span></span>
<span><span class="co">## barcode_counts_aggregated &lt;- barcode_counts_aggregated |&gt;</span></span>
<span><span class="co">##   dplyr::select(!ends_with(".y")) |&gt; </span></span>
<span><span class="co">##   dplyr::rename_with(.fn = ~ gsub("\\.x$", "", .x),</span></span>
<span><span class="co">##                      .cols = dplyr::ends_with(".x"))</span></span>
<span></span>
<span><span class="co">## 3) keep only replicates present in pheno_data</span></span>
<span><span class="va">barcode_counts_aggregated</span> <span class="op">&lt;-</span> <span class="va">barcode_counts_aggregated</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">all_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"barcode_id"</span>, <span class="va">pheno_data</span><span class="op">$</span><span class="va">Replicates_ID</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</figure>
</div>
</div>
<ul>
<li><p>After filtering for background noise, and merging all experiences based on shared barcode IDs, 4629 unique barcode IDs are kept, for 532 samples, see <a href="#lst-background-filtering" class="quarto-xref">Listing&nbsp;<span>2.1</span></a> for details.</p></li>
<li><p><span data-fg="red" style="color: red;">Density plot of controls + justify why + check absence of outliers, to evaluate the relevance of this threshold, <code>HTSFilter</code>, comparison with existing filtering approaches.</span></p></li>
</ul></section><section id="normalisation" class="level3" data-number="2.2.2"><h3 data-number="2.2.2" class="anchored" data-anchor-id="normalisation">
<span class="header-section-number">2.2.2</span> <strong>Normalisation</strong>
</h3>
<ul>
<li>
<span data-fg="blue" style="color: blue;">Normalize barcodes so that the total number of counts per sample is equal to <code>100,000</code></span>. This operation is performed in <a href="#lst-cpm-normalisation" class="quarto-xref">Listing&nbsp;<span>2.2</span></a>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div id="lst-cpm-normalisation" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-cpm-normalisation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.2: Normalise by <span class="math inline">\(100000\)</span>.
</figcaption><div aria-describedby="lst-cpm-normalisation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## 1) normalise by total number of counts ----</span></span>
<span><span class="va">barcode_counts_normalised</span> <span class="op">&lt;-</span> <span class="va">barcode_counts_aggregated</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span></span>
<span>    .cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html">where</a></span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span>,                     </span>
<span>    .fns <span class="op">=</span> <span class="op">~</span> <span class="op">(</span><span class="va">.x</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1e5</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</figure>
</div>
</div>
<ul>
<li><span data-fg="red" style="color: red;">Close to two existing normalisation methods: <strong>Counts Per Million (CPM)</strong> which additionally scales raw counts by total library size and multiplies by <span class="math inline">\(1,000,000\)</span> and <strong>Total Count Scaling (TCS)</strong>: Scales raw counts by the total number of reads (or mapped reads) in each sample, then multiplies by a fixed number (e.g., 100,000). Would compare other normalisation approaches + ignore biological or technical biases + generate MA plots for verifying the mean-variance correction trend + not suitable for DEGs analyses. Apply concatenation phase of all samples before normalising. <a href="https://www.linkedin.com/posts/%F0%9F%8E%AF-ming-tommy-tang-40650014_rna-seq-normalization-what-you-need-to-activity-7313919533780516864-6jFm">Preprocessing and normalisation functions for transcriptomics: a general overview</a></span></li>
</ul></section><section id="prepare-summarizedexperiment-objects-for-normalised-data" class="level3" data-number="2.2.3"><h3 data-number="2.2.3" class="anchored" data-anchor-id="prepare-summarizedexperiment-objects-for-normalised-data">
<span class="header-section-number">2.2.3</span> Prepare <code>SummarizedExperiment</code> objects for normalised data</h3>
<ul>
<li>Higher level of granularity in <a href="#lst-SummarizedExperiment-replicates" class="quarto-xref">Listing&nbsp;<span>2.3</span></a>:</li>
</ul>
<div class="cell" data-layout-align="center">
<div id="lst-SummarizedExperiment-replicates" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-SummarizedExperiment-replicates-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.3: Simply compute the Pearson correlation score at the replicate level, and save the output as a <code>SummarizedExperiment</code>.
</figcaption><div aria-describedby="lst-SummarizedExperiment-replicates-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## 2) prepare SummarizedExperiment inputs ----</span></span>
<span><span class="co">## Barcode counts at replicate level</span></span>
<span><span class="va">barcode_normalised_replicates_mat</span> <span class="op">&lt;-</span> <span class="va">barcode_counts_normalised</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"barcode_id"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Sample metadata</span></span>
<span><span class="va">pheno_data_replicates</span> <span class="op">&lt;-</span> <span class="va">pheno_data</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"Replicates_ID"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">## Feature metadata</span></span>
<span><span class="va">barcode_metadata</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>barcode_id <span class="op">=</span> <span class="va">barcode_counts_normalised</span><span class="op">$</span><span class="va">barcode_id</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">barcode_metadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">barcode_metadata</span><span class="op">$</span><span class="va">barcode_id</span></span>
<span></span>
<span><span class="co">## 3) Buid and Save SummarizedExperiment</span></span>
<span><span class="va">se_replicates</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">barcode_normalised_replicates_mat</span><span class="op">)</span>,</span>
<span>  rowData <span class="op">=</span> <span class="va">barcode_metadata</span>,</span>
<span>  colData <span class="op">=</span> <span class="va">pheno_data_replicates</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</figure>
</div>
</div>
<ul>
<li>Medium level of granularity in <a href="#lst-SummarizedExperiment-compound-per-batch" class="quarto-xref">Listing&nbsp;<span>2.4</span></a>, averaging over replicates:</li>
</ul>
<div class="cell" data-layout-align="center">
<div id="lst-SummarizedExperiment-compound-per-batch" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-SummarizedExperiment-compound-per-batch-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.4: Simply compute the Pearson correlation score at the Compound per Batch level.
</figcaption><div aria-describedby="lst-SummarizedExperiment-compound-per-batch-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">barcode_counts_long</span> <span class="op">&lt;-</span> <span class="va">barcode_counts_normalised</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="va">barcode_id</span>, </span>
<span>                      names_to <span class="op">=</span> <span class="st">"Replicates_ID"</span>,</span>
<span>                      values_to <span class="op">=</span> <span class="st">"Barcode_Counts"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">inner_join</a></span><span class="op">(</span><span class="va">pheno_data</span>, by<span class="op">=</span><span class="st">"Replicates_ID"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">barcode_counts_wider_compound_batch</span> <span class="op">&lt;-</span> <span class="va">barcode_counts_long</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>id_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">barcode_id</span><span class="op">)</span>, </span>
<span>                     names_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">Batch_ID</span>, <span class="va">Compound</span>, <span class="va">Concentrations_ID</span>, <span class="va">Duration_ID</span><span class="op">)</span>,</span>
<span>                     names_sep <span class="op">=</span> <span class="st">":"</span>,</span>
<span>                     values_from <span class="op">=</span> <span class="va">Barcode_Counts</span>, </span>
<span>                     values_fn<span class="op">=</span><span class="va">mean</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pheno_compound_batch</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>original <span class="op">=</span><span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">barcode_counts_wider_compound_batch</span><span class="op">)</span>,<span class="st">"barcode_id"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Batch_ID <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split_i</a></span><span class="op">(</span><span class="va">original</span>, <span class="st">":"</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>         Compound <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split_i</a></span><span class="op">(</span><span class="va">original</span>, <span class="st">":"</span>, <span class="fl">2</span><span class="op">)</span>, </span>
<span>         Concentrations_ID <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split_i</a></span><span class="op">(</span><span class="va">original</span>, <span class="st">":"</span>, <span class="fl">3</span><span class="op">)</span>, </span>
<span>         Duration_ID <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split_i</a></span><span class="op">(</span><span class="va">original</span>, <span class="st">":"</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">inner_join</a></span><span class="op">(</span><span class="va">pheno_data</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Batch_ID</span>, <span class="va">Pathway</span>, <span class="va">MoA</span>, <span class="va">Compound</span>, </span>
<span>                                  <span class="va">Concentrations</span>, <span class="va">Concentrations_ID</span>, <span class="va">Duration</span>, <span class="va">Duration_ID</span><span class="op">)</span>, </span>
<span>             by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Batch_ID"</span>, <span class="st">"Compound"</span>, <span class="st">"Concentrations_ID"</span>, <span class="st">"Duration_ID"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Concentrations_ID</span>, <span class="op">-</span><span class="va">Duration_ID</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"original"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">se_compound_batch</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">barcode_counts_wider_compound_batch</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"barcode_id"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rowData <span class="op">=</span> <span class="va">barcode_metadata</span>,</span>
<span>  colData <span class="op">=</span> <span class="va">pheno_compound_batch</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</figure>
</div>
</div>
<ul>
<li>Low level of granularity in <a href="#lst-SummarizedExperiment-compound" class="quarto-xref">Listing&nbsp;<span>2.5</span></a>, with one CC per compound:</li>
</ul>
<div class="cell" data-layout-align="center">
<div id="lst-SummarizedExperiment-compound" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-SummarizedExperiment-compound-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.5: Compute the Pearson correlation score, after averaging the barcode counts per compound.
</figcaption><div aria-describedby="lst-SummarizedExperiment-compound-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">barcode_counts_wider_by_compound</span> <span class="op">&lt;-</span> <span class="va">barcode_counts_long</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>id_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">barcode_id</span><span class="op">)</span>, </span>
<span>                     names_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">Compound</span><span class="op">)</span>,</span>
<span>                     values_from <span class="op">=</span> <span class="va">Barcode_Counts</span>, </span>
<span>                     values_fn<span class="op">=</span><span class="va">mean</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pheno_compound</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>Compound <span class="op">=</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">barcode_counts_wider_by_compound</span><span class="op">)</span>, <span class="st">"barcode_id"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">inner_join</a></span><span class="op">(</span><span class="va">pheno_data</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Pathway</span>, <span class="va">MoA</span>, <span class="va">Compound</span><span class="op">)</span>, </span>
<span>             by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Compound"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"Compound"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">se_compound</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">barcode_counts_wider_by_compound</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"barcode_id"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rowData <span class="op">=</span> <span class="va">barcode_metadata</span>,</span>
<span>  colData <span class="op">=</span> <span class="va">pheno_compound</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</figure>
</div>
</div>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">se_replicates</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_normalised_per_replicate_"</span>,</span>
<span>                          <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">se_compound_batch</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_normalised_per_compound_by_batch_"</span>,</span>
<span>                          <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">se_compound</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_normalised_per_compound_"</span>,</span>
<span>                          <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section></section><section id="differential-analyses-and-binarisation" class="level2" data-number="2.3"><h2 data-number="2.3" class="anchored" data-anchor-id="differential-analyses-and-binarisation">
<span class="header-section-number">2.3</span> <strong>Differential analyses</strong> and Binarisation</h2>
<ol type="1">
<li><span data-fg="blue" style="color: blue;">Calculate mean of the 4 controls, followed by <strong>Differential Represented Barcode Analysis</strong>: binarise each replicate, assigning a 1 if Fold change is above 3 with respect to Mean value, and 0 otherwise</span></li>
</ol>
<div class="cell" data-layout-align="center">
<div id="lst-differential-barcode-analysis" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-differential-barcode-analysis-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.6: Compute the averaged value for control replicates, then binarise for each compound. A <code>1</code> denoting a significant positive enrichment for a given <code>barcode_id</code>.
</figcaption><div aria-describedby="lst-differential-barcode-analysis-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Step 1: Compute control means per barcode_id and Batch ----</span></span>
<span><span class="va">threshold_FC</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">control_means</span> <span class="op">&lt;-</span> <span class="va">barcode_counts_long</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Compound</span> <span class="op">==</span> <span class="st">"Control"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Batch_ID</span>, <span class="va">barcode_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>Control_Mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">Barcode_Counts</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Step 2: Join control mean to full dataset ----</span></span>
<span><span class="va">barcode_discretised_replicates</span> <span class="op">&lt;-</span> <span class="va">barcode_counts_long</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">control_means</span>, </span>
<span>            by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Batch_ID"</span>,<span class="st">"barcode_id"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Step 3: Compute Fold Change and discretise ----</span></span>
<span><span class="va">barcode_discretised_replicates</span> <span class="op">&lt;-</span> <span class="va">barcode_discretised_replicates</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Compound</span> <span class="op">!=</span> <span class="st">"Control"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Diff <span class="op">=</span> <span class="va">Barcode_Counts</span> <span class="op">-</span> <span class="va">Control_Mean</span>,</span>
<span>    Barcode_Counts <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">Diff</span> <span class="op">&gt;</span> <span class="va">threshold_FC</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co">## Final cleanup </span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Diff</span>, <span class="op">-</span><span class="va">Control_Mean</span><span class="op">)</span>  <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Barcode_Counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/logical.html">as.logical</a></span><span class="op">(</span><span class="va">Barcode_Counts</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</figure>
</div>
</div>
<p><span data-fg="red" style="color: red;">Batch effect correction for integrating across batches? // As Vera emphasised it out, why keeping only positive values? Negative are also interesting. Why not pairing <span class="math inline">\(p\)</span>-values and fold-change (considering indeed really small sample sizes)? // All these operations can be performed with one run, using <code>lm</code>, and a model as such, <span class="math inline">\(\text{Expr} \sim 0 + \text{Gene} + \text{Batch} + \text{Drug}\)</span>, with a <code>contr.treatment</code> design matrix // Perform <strong>sensitivity analyses</strong> to evaluate the impact of threshold criteriaon for binarisation on downstream analyses. <a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></span>.</p>
<section id="summarizedexperiment-of-binarised-markers" class="level3" data-number="2.3.1"><h3 data-number="2.3.1" class="anchored" data-anchor-id="summarizedexperiment-of-binarised-markers">
<span class="header-section-number">2.3.1</span> <code>SummarizedExperiment</code> of binarised markers</h3>
<ul>
<li>
<code>SummarizedExperiment</code> at the replicates level (controls being discarded)</li>
</ul>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Step 4: Save as a SummarizedExperiment ----</span></span>
<span><span class="va">barcode_discretised_replicates_mat</span> <span class="op">&lt;-</span> <span class="va">barcode_discretised_replicates</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>id_cols <span class="op">=</span> <span class="va">barcode_id</span>, names_from <span class="op">=</span> <span class="va">Replicates_ID</span>, values_from <span class="op">=</span> <span class="va">Barcode_Counts</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pheno_replicates_discretised</span> <span class="op">&lt;-</span> <span class="va">pheno_data</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Replicates_ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">barcode_discretised_replicates</span><span class="op">$</span><span class="va">Replicates_ID</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"Replicates_ID"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">se_discretised_replicates</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>binarised <span class="op">=</span> <span class="va">barcode_discretised_replicates_mat</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"barcode_id"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rowData <span class="op">=</span> <span class="va">barcode_metadata</span>,</span>
<span>  colData <span class="op">=</span> <span class="va">pheno_replicates_discretised</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">se_discretised_replicates</span>,</span>
<span>        file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_discretised_per_replicate_"</span>,</span>
<span>                      <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ul>
<li><span data-fg="blue" style="color: blue;">Logical <code>AND</code> over replicates: drug by batch (and concentration and time).</span></li>
</ul>
<div class="cell" data-layout-align="center">
<div id="lst-compound-by-batch-discretised" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-compound-by-batch-discretised-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.7: Average the replicates, using a <code>all</code> operator.
</figcaption><div aria-describedby="lst-compound-by-batch-discretised-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## 2) discretised correlation at the compound by batch level (averaging over replicates), with an all function</span></span>
<span><span class="va">barcode_discretised_compound_batch</span> <span class="op">&lt;-</span> <span class="va">barcode_discretised_replicates</span> <span class="op">|&gt;</span> </span>
<span>   <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>id_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">barcode_id</span><span class="op">)</span>, </span>
<span>                     names_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">Batch_ID</span>, <span class="va">Compound</span>, <span class="va">Concentrations_ID</span>, <span class="va">Duration_ID</span><span class="op">)</span>,</span>
<span>                     names_sep <span class="op">=</span> <span class="st">":"</span>,</span>
<span>                     values_from <span class="op">=</span> <span class="va">Barcode_Counts</span>, </span>
<span>                     values_fn<span class="op">=</span><span class="va">all</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pheno_discretised_compound_batch</span> <span class="op">&lt;-</span> <span class="va">pheno_compound_batch</span><span class="op">[</span><span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">barcode_discretised_compound_batch</span><span class="op">)</span>, <span class="st">"barcode_id"</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">se_compound_batch_discretised</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>discretised <span class="op">=</span> <span class="va">barcode_discretised_compound_batch</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"barcode_id"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rowData <span class="op">=</span> <span class="va">barcode_metadata</span>,</span>
<span>  colData <span class="op">=</span> <span class="va">pheno_discretised_compound_batch</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">se_compound_batch_discretised</span>,</span>
<span>        file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_discretised_per_compound_by_batch_"</span>,</span>
<span>                      <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</figure>
</div>
</div>
<ul>
<li><span data-fg="blue" style="color: blue;">Average over compounds.</span></li>
</ul>
<div class="cell" data-layout-align="center">
<div id="lst-compound-discretised" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-compound-discretised-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2.8: Aggregate at the compound level, assigning a 1 if barcode cell lines were found positiveliy enriched in all compounds.
</figcaption><div aria-describedby="lst-compound-discretised-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## 2) discretised correlation at the compound by batch level (averaging over replicates), with an all function</span></span>
<span><span class="va">barcode_discretised_compound</span> <span class="op">&lt;-</span> <span class="va">barcode_discretised_replicates</span> <span class="op">|&gt;</span> </span>
<span>   <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>id_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">barcode_id</span><span class="op">)</span>, </span>
<span>                     names_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">Compound</span><span class="op">)</span>,</span>
<span>                     names_sep <span class="op">=</span> <span class="st">":"</span>,</span>
<span>                     values_from <span class="op">=</span> <span class="va">Barcode_Counts</span>, </span>
<span>                     values_fn<span class="op">=</span><span class="va">all</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pheno_discretised_compound</span><span class="op">&lt;-</span> <span class="va">pheno_compound</span><span class="op">[</span><span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">barcode_discretised_compound</span><span class="op">)</span>, <span class="st">"barcode_id"</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">se_compound_discretised</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>discretised <span class="op">=</span> <span class="va">barcode_discretised_compound</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"barcode_id"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rowData <span class="op">=</span> <span class="va">barcode_metadata</span>,</span>
<span>  colData <span class="op">=</span> <span class="va">pheno_discretised_compound</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">se_compound_discretised</span>,</span>
<span>        file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_discretised_per_compound_"</span>,</span>
<span>                      <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</figure>
</div>
</div>


</section></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><hr>
<ol>
<li id="fn1"><p>Test the <strong>Helmert contrast</strong> to evaluate <em>compound dose response</em>, or <em>time-course replicates</em>. Also For repeated measures across time, see <a href="https://www.r-bloggers.com/2025/02/one-way-repeated-measure-anova-in-r/">One Way repeated measure ANOVA in R</a><a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/verapancaldilab\.github\.io\/DNABarcode-DrugFingerprint\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        const annoteTargets = window.document.querySelectorAll('.code-annotation-anchor');
        for (let i=0; i<annoteTargets.length; i++) {
          const annoteTarget = annoteTargets[i];
          const targetCell = annoteTarget.getAttribute("data-target-cell");
          const targetAnnotation = annoteTarget.getAttribute("data-target-annotation");
          const contentFn = () => {
            const content = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
            if (content) {
              const tipContent = content.cloneNode(true);
              tipContent.classList.add("code-annotation-tip-content");
              return tipContent.outerHTML;
            }
          }
          const config = {
            allowHTML: true,
            content: contentFn,
            onShow: (instance) => {
              selectCodeLines(instance.reference);
              instance.reference.classList.add('code-annotation-active');
              window.tippy.hideAll();
            },
            onHide: (instance) => {
              unselectCodeLines();
              instance.reference.classList.remove('code-annotation-active');
            },
            maxWidth: 300,
            delay: [50, 0],
            duration: [200, 0],
            offset: [5, 10],
            arrow: true,
            appendTo: function(el) {
              return el.parentElement.parentElement.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'quarto',
            placement: 'right',
            popperOptions: {
              modifiers: [
              {
                name: 'flip',
                options: {
                  flipVariations: false, // true by default
                  allowedAutoPlacements: ['right'],
                  fallbackPlacements: ['right', 'top', 'top-start', 'top-end', 'bottom', 'bottom-start', 'bottom-end', 'left'],
                },
              },
              {
                name: 'preventOverflow',
                options: {
                  mainAxis: false,
                  altAxis: false
                }
              }
              ]        
            }      
          };
          window.tippy(annoteTarget, config); 
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><nav class="page-navigation column-body"><div class="nav-page nav-page-previous">
      <a href="./data-management.html" class="pagination-link" aria-label="Data management: metadata correction and cleansing">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Data management: metadata correction and cleansing</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./results.html" class="pagination-link" aria-label="Results: Heatmaps and Drug networks, based on barcode fingerprints">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Results: Heatmaps and Drug networks, based on barcode fingerprints</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Written by Bastien CHASSAGNOL and Marie-Pier SCOTT-KELLY and Vera PANCALDI and Luca GRUMULATO, with <a href="https://quarto.org/">Quarto</a>.</p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/edit/master/methods.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/blob/master/methods.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link active" href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/" aria-current="page">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


<script src="site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>