<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-GB" xml:lang="en-GB"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>3&nbsp; Results: Heatmaps and Drug networks, based on barcode fingerprints – Inferring the MoA of drugs based on barcode fingerprints.</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./perspectives.html" rel="next">
<link href="./methods.html" rel="prev">
<link href="./assets/favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-86daaaaad7353f9cc0c554efc1dd6d94.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark-f8dc6eab18fde03278982b0b35885446.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-2f278a1226d9ba6da7f61205d60cdf5d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark-64136358cfb85196bf3d352ee32f7b63.min.css" rel="prefetch" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<script async="" src="https://hypothes.is/embed.js"></script><script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script><script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
<meta property="og:title" content="3&nbsp; Results: Heatmaps and Drug networks, based on barcode fingerprints – Inferring the MoA of drugs based on barcode fingerprints.">
<meta property="og:description" content="">
<meta property="og:image" content="https://verapancaldilab.github.io/DNABarcode-DrugFingerprint/results_files/figure-html/fig-heatmap-cell-lines-by-cell-lines-1.png">
<meta property="og:site_name" content="Inferring the MoA of drugs based on barcode fingerprints.">
<meta property="og:image:height" content="960">
<meta property="og:image:width" content="1344">
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./results.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Results: Heatmaps and Drug networks, based on barcode fingerprints</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Cell Barcoding and Drug Fingerprints</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <a href="./Inferring-the-MoA-of-drugs-based-on-barcode-fingerprints..docx" title="Download Docx" class="quarto-navigation-tool px-1" aria-label="Download Docx"><i class="bi bi-file-word"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-management.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Data management: metadata correction and cleansing</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./methods.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Methods: normalise, identify DRBs and save as <code>SummarizedExperiment</code> objects</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./results.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Results: Heatmaps and Drug networks, based on barcode fingerprints</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./perspectives.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Future perspectives: single-cell integration.</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cell-lines-databases.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Appendix: publish to standardised cell line repositories.</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bibliographic References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#reproducibility" id="toc-reproducibility" class="nav-link active" data-scroll-target="#reproducibility"><span class="header-section-number">3.1</span> Reproducibility</a></li>
  <li>
<a href="#compute-compound-similarities" id="toc-compute-compound-similarities" class="nav-link" data-scroll-target="#compute-compound-similarities"><span class="header-section-number">3.2</span> Compute Compound Similarities</a>
  <ul>
<li>
<a href="#compound-heatmaps" id="toc-compound-heatmaps" class="nav-link" data-scroll-target="#compound-heatmaps"><span class="header-section-number">3.2.1</span> Compound Heatmaps</a>
  <ul class="collapse">
<li><a href="#heatmap-with-pearson-correlation-score" id="toc-heatmap-with-pearson-correlation-score" class="nav-link" data-scroll-target="#heatmap-with-pearson-correlation-score"><span class="header-section-number">3.2.1.1</span> Heatmap with Pearson correlation score</a></li>
  <li><a href="#heatmap-with-pearson-correlation-and-binarised-values" id="toc-heatmap-with-pearson-correlation-and-binarised-values" class="nav-link" data-scroll-target="#heatmap-with-pearson-correlation-and-binarised-values"><span class="header-section-number">3.2.1.2</span> Heatmap with Pearson correlation and binarised values</a></li>
  </ul>
</li>
  <li><a href="#compound-networks" id="toc-compound-networks" class="nav-link" data-scroll-target="#compound-networks"><span class="header-section-number">3.2.2</span> Compound Networks</a></li>
  </ul>
</li>
  <li>
<a href="#compute-cell-lines-similarities" id="toc-compute-cell-lines-similarities" class="nav-link" data-scroll-target="#compute-cell-lines-similarities"><span class="header-section-number">3.3</span> Compute Cell Lines Similarities</a>
  <ul>
<li><a href="#cell-line-by-compound" id="toc-cell-line-by-compound" class="nav-link" data-scroll-target="#cell-line-by-compound"><span class="header-section-number">3.3.1</span> Cell Line by Compound</a></li>
  <li><a href="#cell-line-correlation" id="toc-cell-line-correlation" class="nav-link" data-scroll-target="#cell-line-correlation"><span class="header-section-number">3.3.2</span> Cell Line Correlation</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/edit/master/results.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/blob/master/results.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content column-body" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Results: Heatmaps and Drug networks, based on barcode fingerprints</span>
</h1>
</div>



<div class="quarto-title-meta column-body">

    
  
    
  </div>
  


</header><section id="reproducibility" class="level2" data-number="3.1"><h2 data-number="3.1" class="anchored" data-anchor-id="reproducibility">
<span class="header-section-number">3.1</span> Reproducibility</h2>
<p>I list below the R packages required to reproduce the analyses.</p>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## data wrangling</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## reporting</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ardata-fr.github.io/flextable-book/">flextable</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## plotting</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/">igraph</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## auxiliary functions</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="st">"R/utils.R"</span><span class="op">)</span></span>
<span><span class="va">today_date</span> <span class="op">&lt;-</span> <span class="st">"2025-04-28"</span></span>
<span><span class="co">## today_date &lt;- format(Sys.Date(), "%Y-%m-%d")</span></span>
<span></span>
<span><span class="co">## set the seed, for fixing generation of ComplexHeatmaps (dendogram clustering)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section><section id="compute-compound-similarities" class="level2" data-number="3.2"><h2 data-number="3.2" class="anchored" data-anchor-id="compute-compound-similarities">
<span class="header-section-number">3.2</span> Compute Compound Similarities</h2>
<ul>
<li><p><span data-fg="blue" style="color: blue;">Actual computation of the correlation matrix using <code><a href="https://rdrr.io/r/stats/cor.html">stats::cor</a></code>.</span></p></li>
<li><p><span data-fg="red" style="color: red;">Yet, we have to try other correlation methods on the continuous space, or consider other metrics if working on the discrete space. Unbalanced Wassertein distance, or relatives, for distinct input and output dimensions.</span></p></li>
</ul>
<section id="compound-heatmaps" class="level3" data-number="3.2.1"><h3 data-number="3.2.1" class="anchored" data-anchor-id="compound-heatmaps">
<span class="header-section-number">3.2.1</span> Compound Heatmaps</h3>
<ul>
<li>
<p>Some tutorials:</p>
<ul>
<li><p><a href="https://www.linkedin.com/posts/%F0%9F%8E%AF-ming-tommy-tang-40650014_making-a-heatmap-is-an-essential-skill-for-activity-7317543424511868928-F25t">Blend of useful resources on generating Heatmaps</a>:</p></li>
<li><p><a href="https://www.nature.com/articles/nmeth.2134">Mapping quantitative data to color</a>, from <span class="citation" data-cites="gehlenborg2012nm">Gehlenborg and Wong (<a href="references.html#ref-gehlenborg2012nm" role="doc-biblioref">2012</a>)</span>:</p></li>
<li>
<p><strong>Real-Life Examples</strong> with <code>ComplexHeatmap</code>:</p>
<ul>
<li><a href="https://rpubs.com/crazyhottommy/heatmap_demystified">Heatmap with transcriptomic expression</a></li>
<li><a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/more-examples.html">Heatmap with genomic expression</a></li>
<li><a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/genome-level-heatmap.html">Genome-level Heatmap</a></li>
<li>
<a href="https://gdevailly.netlify.app/post/plotting-big-matrices-in-r">Plotting large heatmaps in R</a> and <a href="https://jokergoo.github.io/2020/06/30/rasterization-in-complexheatmap"><em>Rasterisation</em>, with thousands of genes to be considered</a>.</li>
<li><a href="https://jokergoo.github.io/2020/05/11/set-cell-width/height-in-the-heatmap/">Control size of Heatmaps</a></li>
</ul>
</li>
<li><p><strong>Real-life examples</strong> with <a href="https://github.com/theislab/atlas-feature-selection-benchmark/blob/b89fc0f66747062e6e1b4b35bd392b27ad035295/analysis/R/plotting.R#L8"><code>ggplot2</code></a></p></li>
<li><p><strong>Real-Life examples</strong> with <a href="https://funkyheatmap.github.io/"><code>funkyheatmap</code> CRAN package</a></p></li>
</ul>
</li>
</ul>
<section id="heatmap-with-pearson-correlation-score" class="level4" data-number="3.2.1.1"><h4 data-number="3.2.1.1" class="anchored" data-anchor-id="heatmap-with-pearson-correlation-score">
<span class="header-section-number">3.2.1.1</span> Heatmap with Pearson correlation score</h4>
<ul>
<li>Heatmap in <span class="quarto-unresolved-ref">?fig-heatmap-normalised-replicate</span> is generated at the replicate level:</li>
</ul>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Read SummarizedExperiment object</span></span>
<span><span class="va">se_normalised_replicate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_normalised_per_replicate_"</span>, <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_normalised_replicate_mat</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_normalised_replicate</span><span class="op">)</span></span>
<span><span class="va">se_normalised_pheno_data_replicate</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_normalised_replicate</span><span class="op">)</span></span>
<span><span class="va">se_normalised_pheno_data_replicate</span><span class="op">$</span><span class="va">log10Conc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">se_normalised_pheno_data_replicate</span><span class="op">$</span><span class="va">Concentrations</span><span class="op">==</span><span class="fl">0</span>, </span>
<span>                                   <span class="fl">0</span>, </span>
<span>                                   <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">se_normalised_pheno_data_replicate</span><span class="op">$</span><span class="va">Concentrations</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Compute global correlation score</span></span>
<span><span class="va">cor_per_replicate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">se_normalised_replicate_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define clustering colours</span></span>
<span><span class="va">col_heatmap_scale</span> <span class="op">&lt;-</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">cor_per_replicate</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define custom colour Heatmap annotation</span></span>
<span><span class="co">## num_batches &lt;- length(unique(se_normalised_pheno_data_replicate$Batch_ID))</span></span>
<span><span class="co">## num_compounds &lt;- length(unique(se_normalised_pheno_data_replicate$Compound))</span></span>
<span><span class="co">## num_moa &lt;- length(unique(se_normalised_pheno_data_replicate$MoA))</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## batch_col &lt;- setNames(colorRampPalette(brewer.pal(8, "Dark2"))(num_batches), </span></span>
<span><span class="co">##                       unique(se_normalised_pheno_data_replicate$Batch_ID))</span></span>
<span><span class="co">## compound_col &lt;- setNames(colorRampPalette(brewer.pal(11, "Spectral"))(num_compounds), </span></span>
<span><span class="co">##                          unique(se_normalised_pheno_data_replicate$Compound))</span></span>
<span><span class="co">## moa_col &lt;- setNames(colorRampPalette(brewer.pal(12, "Set3"))(num_moa), </span></span>
<span><span class="co">##                     unique(se_normalised_pheno_data_replicate$MoA))</span></span>
<span></span>
<span></span>
<span><span class="co">## Heatmap annotation</span></span>
<span><span class="va">top_ha</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">se_normalised_pheno_data_replicate</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Batch_ID</span>, <span class="va">MoA</span>, <span class="va">Compound</span><span class="op">)</span><span class="op">)</span> </span>
<span>                                        <span class="co">## col = list(Batch_ID = batch_col, </span></span>
<span>                                        <span class="co">##            MoA = moa_col, </span></span>
<span>                                        <span class="co">##            Compound = compound_col))</span></span>
<span></span>
<span><span class="va">row_ha</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/rowAnnotation.html">rowAnnotation</a></span><span class="op">(</span>Concentration <span class="op">=</span></span>
<span>                                          <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/anno_barplot.html">anno_barplot</a></span><span class="op">(</span><span class="va">se_normalised_pheno_data_replicate</span><span class="op">$</span><span class="va">log10Conc</span><span class="op">)</span>, </span>
<span>                                        Duration <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">se_normalised_pheno_data_replicate</span><span class="op">$</span><span class="va">Duration</span>, <span class="st">"d"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Generate Heatmap</span></span>
<span><span class="va">heatmap_normalised_replicates</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">cor_per_replicate</span>, </span>
<span>                                          col<span class="op">=</span><span class="va">col_heatmap_scale</span>, </span>
<span>                                          name <span class="op">=</span> <span class="st">"CC for all Replicates"</span>, </span>
<span>                                          show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          show_column_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                          row_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          column_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                          row_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, </span>
<span>                                          top_annotation <span class="op">=</span> <span class="va">top_ha</span>, </span>
<span>                                          right_annotation <span class="op">=</span> <span class="va">row_ha</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html">grid.grabExpr</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"figures/compounds/heatmap_normalised_replicates_"</span>, <span class="va">today_date</span>,<span class="st">".pdf"</span><span class="op">)</span>, </span>
<span>       <span class="va">heatmap_normalised_replicates</span>, dpi <span class="op">=</span> <span class="fl">600</span>, width <span class="op">=</span> <span class="fl">20</span>, height <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ul>
<li>Heatmap in <span class="quarto-unresolved-ref">?fig-heatmap-normalised-compound-batch</span> is generated by averaging normalised barcode counts over replicates:</li>
</ul>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Read SummarizedExperiment object</span></span>
<span><span class="va">se_normalised_compound_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_normalised_per_compound_by_batch_"</span>, <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_normalised_compound_batch_mat</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_normalised_compound_batch</span><span class="op">)</span></span>
<span><span class="va">se_pheno_data_normalised_compound_batch</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_normalised_compound_batch</span><span class="op">)</span></span>
<span><span class="co">## Compute global correlation score</span></span>
<span><span class="va">cor_per_compound_by_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">se_normalised_compound_batch_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define clustering colours</span></span>
<span><span class="va">col_heatmap_scale</span> <span class="op">&lt;-</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">cor_per_compound_by_batch</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Heatmap annotation</span></span>
<span><span class="va">ha</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">se_pheno_data_normalised_compound_batch</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Pathway</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define Heatmap</span></span>
<span><span class="va">heatmap_normalised_compound_batch</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">cor_per_compound_by_batch</span>, </span>
<span>                                          col<span class="op">=</span><span class="va">col_heatmap_scale</span>, </span>
<span>                                          name <span class="op">=</span> <span class="st">"CC for all Compound"</span>, </span>
<span>                                          show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          show_column_names <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                          row_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          column_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                          column_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>                                          row_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>, </span>
<span>                                          top_annotation <span class="op">=</span> <span class="va">ha</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html">grid.grabExpr</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"figures/compounds/heatmap_normalised_compound_by_batch_"</span>, <span class="va">today_date</span>,<span class="st">".pdf"</span><span class="op">)</span>, </span>
<span>       <span class="va">heatmap_normalised_compound_batch</span>, dpi <span class="op">=</span> <span class="fl">600</span>, width <span class="op">=</span> <span class="fl">20</span>, height <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ul>
<li>Heatmap in <span class="quarto-unresolved-ref">?fig-heatmap-global-compound</span> is generated at the compound level:</li>
</ul>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Read SummarizedExperiment object</span></span>
<span><span class="va">se_normalised_compound</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_normalised_per_compound_"</span>, <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_normalised_compound_mat</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_normalised_compound</span><span class="op">)</span></span>
<span><span class="va">se_normalised_pheno_data_compound</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_normalised_compound</span><span class="op">)</span></span>
<span><span class="co">## Compute global correlation score</span></span>
<span><span class="va">cor_per_compound</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">se_normalised_compound_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define clustering colours</span></span>
<span><span class="va">col_heatmap_scale</span> <span class="op">&lt;-</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">cor_per_compound</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Heatmap annotation</span></span>
<span><span class="va">ha</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">se_normalised_pheno_data_compound</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Pathway</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define Heatmap</span></span>
<span><span class="va">heatmap_normalised_compound</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">cor_per_compound</span>, </span>
<span>                                          col<span class="op">=</span><span class="va">col_heatmap_scale</span>, </span>
<span>                                          name <span class="op">=</span> <span class="st">"CC for all Compound"</span>, </span>
<span>                                          show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          show_column_names <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                          row_dend_reorder <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                          column_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                          <span class="co">## font size</span></span>
<span>                                          row_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>, </span>
<span>                                          top_annotation <span class="op">=</span> <span class="va">ha</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html">grid.grabExpr</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"figures/compounds/heatmap_normalised_compound_"</span>, <span class="va">today_date</span>,<span class="st">".pdf"</span><span class="op">)</span>, </span>
<span>       <span class="va">heatmap_normalised_compound</span>, dpi <span class="op">=</span> <span class="fl">600</span>, width <span class="op">=</span> <span class="fl">20</span>, height <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section><section id="heatmap-with-pearson-correlation-and-binarised-values" class="level4" data-number="3.2.1.2"><h4 data-number="3.2.1.2" class="anchored" data-anchor-id="heatmap-with-pearson-correlation-and-binarised-values">
<span class="header-section-number">3.2.1.2</span> Heatmap with Pearson correlation and binarised values</h4>
<ul>
<li>Heatmap at the replicates level in <span class="quarto-unresolved-ref">?fig-heatmap-binarised-replicates</span>:</li>
</ul>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Read SummarizedExperiment object</span></span>
<span><span class="va">se_binarised_replicates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_discretised_per_replicate_"</span>, <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_binarised_replicates_mat</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_binarised_replicates</span><span class="op">)</span></span>
<span><span class="va">se_binarised_pheno_data_replicate</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_binarised_replicates</span><span class="op">)</span></span>
<span><span class="co">## Compute global correlation score</span></span>
<span><span class="va">cor_per_replicates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">se_binarised_replicates_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define clustering colours</span></span>
<span><span class="va">col_heatmap_scale</span> <span class="op">&lt;-</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">cor_per_replicates</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Heatmap annotation</span></span>
<span><span class="va">ha</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">se_binarised_pheno_data_replicate</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Pathway</span>, <span class="va">Batch_ID</span>, <span class="va">Compound</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">heatmap_binarised_replicates</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">cor_per_replicates</span>, </span>
<span>                                            col<span class="op">=</span><span class="va">col_heatmap_scale</span>, </span>
<span>                                            name <span class="op">=</span> <span class="st">"CC at binarised Compound Level"</span>,</span>
<span>                                            show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                            show_column_names <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                            row_dend_reorder <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                            column_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                            <span class="co">## font size</span></span>
<span>                                            row_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, </span>
<span>                                            top_annotation <span class="op">=</span> <span class="va">ha</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html">grid.grabExpr</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"figures/compounds/heatmap_binarised_replicates_"</span>, <span class="va">today_date</span>,<span class="st">".pdf"</span><span class="op">)</span>, </span>
<span>       <span class="va">heatmap_binarised_replicates</span>, dpi <span class="op">=</span> <span class="fl">600</span>, width <span class="op">=</span> <span class="fl">20</span>, height <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>                                            </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ul>
<li>Heatmap at the compound by batch level in <span class="quarto-unresolved-ref">?fig-heatmap-binarised-compound-by-batch</span> (aggregated over technical replicates):</li>
</ul>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Read SummarizedExperiment object</span></span>
<span><span class="va">se_binarised_compound_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_discretised_per_compound_by_batch_"</span>, <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_binarised_compound_batch_mat</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_binarised_compound_batch</span><span class="op">)</span></span>
<span><span class="va">se_binarised_pheno_data_compound_batch</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_binarised_compound_batch</span><span class="op">)</span></span>
<span><span class="co">## Compute global correlation score</span></span>
<span><span class="va">cor_per_compound_by_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">se_binarised_compound_batch_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define clustering colours</span></span>
<span><span class="va">col_heatmap_scale</span> <span class="op">&lt;-</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">cor_per_compound_by_batch</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Heatmap annotation</span></span>
<span><span class="va">ha</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">se_binarised_pheno_data_compound_batch</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Pathway</span>, <span class="va">MoA</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">heatmap_binarised_compound_batch</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">cor_per_compound_by_batch</span>, </span>
<span>                                            col<span class="op">=</span><span class="va">col_heatmap_scale</span>, </span>
<span>                                            name <span class="op">=</span> <span class="st">"CC at binarised Compound Level"</span>,</span>
<span>                                            show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                            show_column_names <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                            row_dend_reorder <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                            column_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                            <span class="co">## font size</span></span>
<span>                                            row_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>, </span>
<span>                                            top_annotation <span class="op">=</span> <span class="va">ha</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html">grid.grabExpr</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"figures/compounds/heatmap_binarised_compound_by_batch_"</span>, <span class="va">today_date</span>,<span class="st">".pdf"</span><span class="op">)</span>, </span>
<span>       <span class="va">heatmap_binarised_compound_batch</span>, dpi <span class="op">=</span> <span class="fl">600</span>, width <span class="op">=</span> <span class="fl">20</span>, height <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>                                            </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ul>
<li>Heatmap at the compound level in <span class="quarto-unresolved-ref">?fig-heatmap-binarised-compound</span>.</li>
</ul>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Read SummarizedExperiment object</span></span>
<span><span class="va">se_binarised_compound</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/se_discretised_per_compound_"</span>, <span class="va">today_date</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_binarised_compound_mat</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">se_binarised_compound</span><span class="op">)</span></span>
<span><span class="va">se_binarised_pheno_data_compound</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se_binarised_compound</span><span class="op">)</span></span>
<span><span class="co">## Compute global correlation score</span></span>
<span><span class="va">cor_per_compound</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">se_binarised_compound_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define clustering colours</span></span>
<span><span class="va">col_heatmap_scale</span> <span class="op">&lt;-</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">cor_per_compound</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Heatmap annotation</span></span>
<span><span class="va">ha</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">se_binarised_pheno_data_compound</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Pathway</span>, <span class="va">MoA</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">heatmap_binarised_compound</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">cor_per_compound</span>, </span>
<span>                                            col<span class="op">=</span><span class="va">col_heatmap_scale</span>, </span>
<span>                                            name <span class="op">=</span> <span class="st">"CC at binarised Compound Level"</span>,</span>
<span>                                            show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                            show_column_names <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                            column_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>                                            row_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                            column_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                            row_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>, </span>
<span>                                            top_annotation <span class="op">=</span> <span class="va">ha</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html">grid.grabExpr</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"figures/compounds/heatmap_binarised_compound_"</span>, <span class="va">today_date</span>,<span class="st">".pdf"</span><span class="op">)</span>, </span>
<span>       <span class="va">heatmap_binarised_compound</span>, dpi <span class="op">=</span> <span class="fl">600</span>, width <span class="op">=</span> <span class="fl">20</span>, height <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>                                            </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<!-- ```{r} -->
<!-- #| label: combine-Heatmap -->
<!-- p1 <- grid.grabExpr(ComplexHeatmap::draw(heatmap_compounds)) -->
<!-- p2 <- grid.grabExpr(ComplexHeatmap::draw(heatmap_replicates)) -->
<!-- heatmaps_global <- plot_grid(p1, p2, nrow = 1, labels = c("A", "B")) -->
<!-- ggsave("figures/heatmap_fingerprints.pdf",  -->
<!--        heatmaps_global, dpi = 600) -->
<!-- ``` -->
</section></section><section id="compound-networks" class="level3" data-number="3.2.2"><h3 data-number="3.2.2" class="anchored" data-anchor-id="compound-networks">
<span class="header-section-number">3.2.2</span> Compound Networks</h3>
<ul>
<li>Plot weighted undirected compound graphs with <code><a href="https://r.igraph.org/reference/graph_from_adjacency_matrix.html">igraph::graph_from_adjacency_matrix</a></code> in <span class="quarto-unresolved-ref">?fig-igraph-drug</span>, filtering out pairwise connections between 2 drugs if the <span class="math inline">\(p\)</span>-value is not significant and absolute correlation score is below 0.8:</li>
</ul>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## 1) Compute correlation matrix and p-values ----</span></span>
<span><span class="va">cor_res</span> <span class="op">&lt;-</span> <span class="fu">Hmisc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Hmisc/man/rcorr.html">rcorr</a></span><span class="op">(</span><span class="va">se_binarised_compound_mat</span>, type <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span></span>
<span><span class="va">cor_binarised_compound</span> <span class="op">&lt;-</span> <span class="va">cor_res</span><span class="op">$</span><span class="va">r</span></span>
<span><span class="va">pval_binarised_compound</span> <span class="op">&lt;-</span> <span class="va">cor_res</span><span class="op">$</span><span class="va">P</span></span>
<span></span>
<span><span class="va">adj_mat</span> <span class="op">&lt;-</span> <span class="va">cor_binarised_compound</span></span>
<span><span class="va">adj_mat</span><span class="op">[</span><span class="va">pval_binarised_compound</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">cor_res</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co">## 2) build igraph adjacency matrix ----</span></span>
<span><span class="va">compound_graph</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/graph_from_adjacency_matrix.html">graph_from_adjacency_matrix</a></span><span class="op">(</span>adjmatrix<span class="op">=</span><span class="va">adj_mat</span>, </span>
<span>                                          weighted <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          diag <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                          mode <span class="op">=</span> <span class="st">"undirected"</span><span class="op">)</span></span>
<span><span class="co">## 3) igraph customisation ----</span></span>
<span><span class="va">compound_graph</span><span class="op">$</span><span class="va">date</span> <span class="op">&lt;-</span> <span class="va">today_date</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/E.html">E</a></span><span class="op">(</span><span class="va">compound_graph</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/E.html">E</a></span><span class="op">(</span><span class="va">compound_graph</span><span class="op">)</span><span class="op">$</span><span class="va">weight</span> <span class="op">&gt;</span><span class="fl">0</span>, <span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/E.html">E</a></span><span class="op">(</span><span class="va">compound_graph</span><span class="op">)</span><span class="op">$</span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/E.html">E</a></span><span class="op">(</span><span class="va">compound_graph</span><span class="op">)</span><span class="op">$</span><span class="va">weight</span><span class="op">)</span></span>
<span></span>
<span><span class="va">layout_graph</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/layout_with_fr.html">layout_with_fr</a></span><span class="op">(</span><span class="va">compound_graph</span>,</span>
<span>                                       niter <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vertex_structure</span> <span class="op">&lt;-</span> <span class="va">se_binarised_pheno_data_compound</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">Pathway</span><span class="op">==</span><span class="st">"EGFR"</span>, <span class="st">"green"</span>, <span class="st">"yellow"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/V.html">V</a></span><span class="op">(</span><span class="va">compound_graph</span><span class="op">)</span><span class="op">$</span><span class="va">color</span> <span class="op">&lt;-</span> <span class="va">vertex_structure</span><span class="op">$</span><span class="va">color</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"figures/compounds/compound_graph_"</span>, <span class="va">today_date</span>, <span class="st">".pdf"</span><span class="op">)</span>, </span>
<span>    width <span class="op">=</span> <span class="fl">16</span>, height <span class="op">=</span> <span class="fl">16</span>, useDingbats <span class="op">=</span> <span class="cn">TRUE</span>, compress <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compound_graph</span>, layout <span class="op">=</span> <span class="va">layout_graph</span>, vertex.size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>     <span class="co">## vertex.label.dist= 2,</span></span>
<span>     main <span class="op">=</span> <span class="st">"Compound network interaction.</span></span>
<span><span class="st">     In green, drugs related with the EGFR mechanism,</span></span>
<span><span class="st">     and in yellow other pathways.</span></span>
<span><span class="st">     Edge widths are proportional to the coefficient of correlation."</span>, </span>
<span>     edge.width <span class="op">=</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/E.html">E</a></span><span class="op">(</span><span class="va">compound_graph</span><span class="op">)</span><span class="op">$</span><span class="va">weight</span> <span class="op">*</span> <span class="fl">20</span>,</span>
<span>     vertex.label.cex <span class="op">=</span> <span class="fl">1</span>, curved <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  png </span></span>
<span><span class="co">##    2</span></span>
<span></span>
<span><span class="co">## 4) export PNG and readable graph formats for Neo4J</span></span>
<span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/write_graph.html">write_graph</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">compound_graph</span>, </span>
<span>                    file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"./results/compounds/compound_network_"</span>, <span class="va">today_date</span>, <span class="st">".graphml"</span><span class="op">)</span>,</span>
<span>                    format <span class="op">=</span> <span class="st">"graphml"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section></section><section id="compute-cell-lines-similarities" class="level2" data-number="3.3"><h2 data-number="3.3" class="anchored" data-anchor-id="compute-cell-lines-similarities">
<span class="header-section-number">3.3</span> Compute Cell Lines Similarities</h2>
<section id="cell-line-by-compound" class="level3" data-number="3.3.1"><h3 data-number="3.3.1" class="anchored" data-anchor-id="cell-line-by-compound">
<span class="header-section-number">3.3.1</span> Cell Line by Compound</h3>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">se_binarised_compound_batch_fac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">se_binarised_compound_batch_mat</span>, <span class="fl">2</span>, <span class="va">as.character</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Heatmap annotation</span></span>
<span><span class="va">ha</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">se_binarised_pheno_data_compound_batch</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                                          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Batch_ID</span>, <span class="va">Pathway</span>, <span class="va">Compound</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">heatmap_cell_lines_by_sample</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">se_binarised_compound_batch_fac</span>, </span>
<span>                                                        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span>, </span>
<span>                                                        name <span class="op">=</span> <span class="st">"Heatmap: cell-line by sample"</span>,</span>
<span>                                                        show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                                        row_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                        <span class="co">## clustering_distance_rows = "pearson",</span></span>
<span>                                                        show_row_dend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                                        column_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                                        show_column_names <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                                        <span class="co">## clustering_distance_columns = "pearson",</span></span>
<span>                                                        column_names_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>, </span>
<span>                                                        top_annotation <span class="op">=</span> <span class="va">ha</span>, </span>
<span>                                                        <span class="co">## compression issues</span></span>
<span>                                                        use_raster <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html">grid.grabExpr</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"figures/cell_lines/heatmap_cell_lines_by_samples_"</span>, <span class="va">today_date</span>,<span class="st">".pdf"</span><span class="op">)</span>, </span>
<span>       <span class="va">heatmap_cell_lines_by_sample</span>, dpi <span class="op">=</span> <span class="fl">300</span>, width <span class="op">=</span> <span class="fl">20</span>, height <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>                                            </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section><section id="cell-line-correlation" class="level3" data-number="3.3.2"><h3 data-number="3.3.2" class="anchored" data-anchor-id="cell-line-correlation">
<span class="header-section-number">3.3.2</span> Cell Line Correlation</h3>
<div class="cell" data-layout-align="center">
<details open="" class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## 1) evaluation of cell resistance</span></span>
<span><span class="va">present_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">se_binarised_compound_batch_mat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">present_cells</span>, </span>
<span>     breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">present_cells</span><span class="op">)</span><span class="op">)</span>,</span>
<span>     main<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 2) filtered Heatmaps </span></span>
<span><span class="va">threshold_cell_prevalence</span> <span class="op">&lt;-</span> <span class="fl">10</span> <span class="co">## alternatively, remove cells that are not present in all samples</span></span>
<span><span class="va">se_binarised_compact</span> <span class="op">&lt;-</span> <span class="va">se_binarised_compound_batch_mat</span><span class="op">[</span><span class="va">present_cells</span> <span class="op">&gt;</span></span>
<span>                                                             <span class="va">threshold_cell_prevalence</span>, <span class="op">]</span></span>
<span><span class="va">cor_cell_lines</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">se_binarised_compact</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">col_heatmap_scale</span> <span class="op">&lt;-</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">cor_cell_lines</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                                            <span class="fl">0</span>, </span>
<span>                                            <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">cor_cell_lines</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 3) generate Heatmap of correlation cell lines</span></span>
<span><span class="va">heatmap_cell_lines_by_cell_lines</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">cor_cell_lines</span>, </span>
<span>                                                        col <span class="op">=</span> <span class="va">col_heatmap_scale</span>, </span>
<span>                                                        name <span class="op">=</span> <span class="st">"Heatmap: cell-line by cell line"</span>,</span>
<span>                                                        show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                                        row_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                        show_row_dend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                        column_dend_reorder <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                                        show_column_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                                        na_col <span class="op">=</span> <span class="st">"black"</span>, </span>
<span>                                                        use_raster <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                                        raster_resize_mat <span class="op">=</span> <span class="va">mean</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html">grid.grabExpr</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"figures/cell_lines/heatmap_cell_lines_by_cell_lines_"</span>, <span class="va">today_date</span>,<span class="st">".pdf"</span><span class="op">)</span>, </span>
<span>       <span class="va">heatmap_cell_lines_by_cell_lines</span>, dpi <span class="op">=</span> <span class="fl">600</span>, width <span class="op">=</span> <span class="fl">20</span>, height <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<div id="fig-heatmap-cell-lines-by-cell-lines" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-heatmap-cell-lines-by-cell-lines-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="results_files/figure-html/fig-heatmap-cell-lines-by-cell-lines-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="Figure&nbsp;3.1: Histogram of the distribution of cell resistance (for a given barcode ID, returns the total number of cell lines identified)."><img src="results_files/figure-html/fig-heatmap-cell-lines-by-cell-lines-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-heatmap-cell-lines-by-cell-lines-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.1: Histogram of the distribution of cell resistance (for a given barcode ID, returns the total number of cell lines identified).
</figcaption></figure>
</div>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-gehlenborg2012nm" class="csl-entry" role="listitem">
Gehlenborg, Nils, and Bang Wong. 2012. <span>‘Mapping Quantitative Data to Color’</span>. <em>Nature Methods</em> 9 (8): 769–69. <a href="https://doi.org/10.1038/nmeth.2134">https://doi.org/10.1038/nmeth.2134</a>.
</div>
</div>
</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/verapancaldilab\.github\.io\/DNABarcode-DrugFingerprint\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      const annoteTargets = window.document.querySelectorAll('.code-annotation-anchor');
      for (let i=0; i<annoteTargets.length; i++) {
        const annoteTarget = annoteTargets[i];
        const targetCell = annoteTarget.getAttribute("data-target-cell");
        const targetAnnotation = annoteTarget.getAttribute("data-target-annotation");
        const contentFn = () => {
          const content = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          if (content) {
            const tipContent = content.cloneNode(true);
            tipContent.classList.add("code-annotation-tip-content");
            return tipContent.outerHTML;
          }
        }
        const config = {
          allowHTML: true,
          content: contentFn,
          onShow: (instance) => {
            selectCodeLines(instance.reference);
            instance.reference.classList.add('code-annotation-active');
            window.tippy.hideAll();
          },
          onHide: (instance) => {
            unselectCodeLines();
            instance.reference.classList.remove('code-annotation-active');
          },
          maxWidth: 300,
          delay: [50, 0],
          duration: [200, 0],
          offset: [5, 10],
          arrow: true,
          appendTo: function(el) {
            return el.parentElement.parentElement.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'quarto',
          placement: 'right',
          popperOptions: {
            modifiers: [
            {
              name: 'flip',
              options: {
                flipVariations: false, // true by default
                allowedAutoPlacements: ['right'],
                fallbackPlacements: ['right', 'top', 'top-start', 'top-end', 'bottom', 'bottom-start', 'bottom-end', 'left'],
              },
            },
            {
              name: 'preventOverflow',
              options: {
                mainAxis: false,
                altAxis: false
              }
            }
            ]        
          }      
        };
        window.tippy(annoteTarget, config); 
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation column-body"><div class="nav-page nav-page-previous">
      <a href="./methods.html" class="pagination-link" aria-label="Methods: normalise, identify DRBs and save as `SummarizedExperiment` objects">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Methods: normalise, identify DRBs and save as <code>SummarizedExperiment</code> objects</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./perspectives.html" class="pagination-link" aria-label="Future perspectives: single-cell integration.">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Future perspectives: single-cell integration.</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Written by Bastien CHASSAGNOL and Marie-Pier SCOTT-KELLY and Vera PANCALDI and Luca GRUMULATO, with <a href="https://quarto.org/">Quarto</a>.</p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/edit/master/results.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/blob/master/results.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link active" href="https://github.com/VeraPancaldiLab/DNABarcode-DrugFingerprint/" aria-current="page">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer><script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>


<script src="site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>